### PR TITLE
Add async checksum generation for resource files

### PR DESCRIPTION
**Pull request by Pirulax**
Basically, resource file checksums were checked on the main thread(problem one), also, they were checked, in worst case scenario, 3 times(on the first load), and best case scenario 2.

This PR adds async checksum generation, as well as removes redundant checks. It works perfectly for me, I had no issues whatsoever with it. 

Some notes to who ever tests this PR:
- `Packet_StartResource` is called every time a resources needs to be started. No matter whenever its started with a command / is in `mtaconfig`

- After a resource is stopped all its associated `CResourceFiles` are deleted, which means, after restarting the resource their checksums are rechecked

- Adding the additional `const bool bCanDownload` was necessary to avoid a 'download loop' [as i call it], which occurs when i dont know how / when, but if it does, the resource is stuck in a loop, and tries to download the file over and over again

- `bIsSehrGut` is serious, and without it the code would explode

- Im sorry for the 'addendums', but I just totally forgot about a few things, but I wanted to keep the commits as clean as possible. 

- A rebase will be needed to remove debug stuff

Edit:
Bugs discovered:
- Seems like after updating a resource file, it doesnt re-download it. Needs further investigation
- Make sure there's enough RAM to load the whole file into memory, if not use the fallback, old method(65 kb stack buffer)

### CODE REVIEW

File: Client/mods/deathmatch/logic/CResource.cpp [hunk 1]
```diff
@@ -226,35 +226,127 @@ bool CResource::CallExportedFunction(const char* szFunctionName, CLuaArguments&
...
+    }
+
+    // Make sure all checksums are generated
+    // We do this here, so it can run async
+    for (auto* pFile : m_ResourceFiles)
+        if (pFile->HasClientChecksumGenerated())
+            pFile->GenerateClientChecksumAsync();
+
+    // Files should be downloaded by now, let's check their hashes, and re-download if needed
+    bool bAllWasSehrGut = true;
```
Comment (Lpsd) [inline Client/mods/deathmatch/logic/CResource.cpp:277]: Can you use an English var name?

Comment (Pirulax) [reply Client/mods/deathmatch/logic/CResource.cpp:277]: :(
It was meant to be a joke. 
But sure, if it bothers anyone I can change it, no problem.

File: Client/mods/deathmatch/logic/CDownloadableResource.h [hunk 2]
```diff
@@ -20,11 +20,12 @@
 
 #include <bochs_internal/crc32.h>
 #include "CChecksum.h"
+#include <future>
 
 class CDownloadableResource
 {
 public:
-    enum eResourceType
+    enum eResourceType : unsigned char // Must be unsigned char, so in CClientGame::Packet_StartResource it's read properly from the packet.
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/CDownloadableResource.h:28]: put the comment above on a new line above

File: Client/mods/deathmatch/logic/CDownloadableResource.cpp [hunk 3]
```diff
@@ -10,23 +10,20 @@
  *****************************************************************************/
 
 #include <StdInc.h>
+#include <future>     // return value from SharedUtil::async
+
+#define INVALIDATE_OLD_FUTURE if (m_GenerateChecksumFuture.valid()) m_GenerateChecksumFuture = std::future<CChecksum>();
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/CDownloadableResource.cpp:15]: what's wrong with a private function?

Comment (Pirulax) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:15]: I felt like this is neater, than declaring a 4-5 line function.

Comment (qaisjp) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:15]: macros are undesirable

Comment (ghost) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:15]: > macros are undesirable

agreed

Comment (Pirulax) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:15]: agreed

**Pull request closed without merging**